### PR TITLE
Add SPI example using embedded_hal_bus

### PR DESCRIPTION
**Pull request by 9names**

### CODE REVIEW

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+/// We need to use a generic here (SPI), because we want our driver to work for any other
+/// microcontroller that implements the embedded-hal Spi traits
+struct MySpiDriver<SPI> {
+    spi: SPI,
+}
+
+/// When things go wrong, we could return Error(()) but that wouldn't help anyone troubleshoot
+/// so we'll create an error type with additional info to return.
+#[derive(Copy, Clone, Debug)]
+enum MyError<SPI> {
```
Comment (thejpster) [inline rp2040-hal-examples/src/bin/spi_eh_bus.rs:59]: Maybe call this type parameters or ERR?

File: rp2040-hal-examples/Cargo.toml [hunk 2]
```diff
@@ -32,5 +33,8 @@ panic-halt = "0.2.0"
 panic-probe = {version = "0.3.1", features = ["print-defmt"]}
 pio = "0.2.0"
 pio-proc = "0.2.0"
+# We aren't using this, but embedded-hal-bus 0.2 unconditionally requires atomics.
+# Should be fixed in e-h-b 0.3 via https://github.com/rust-embedded/embedded-hal/pull/607
+portable-atomic = { version = "1.7.0", features = ["critical-section"] }
```
Comment (thejpster) [inline rp2040-hal-examples/Cargo.toml:38]: The toml formatter doesn't like spaces inside the braces

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+///
+/// The `#[rp2040_hal::entry]` macro ensures the Cortex-M start-up code calls this function
+/// as soon as all global variables and the spinlock are initialised.
+///
+/// The function configures the RP2040 peripherals, then performs some example
+/// SPI transactions, then goes to sleep.
+#[rp2040_hal::entry]
+fn main() -> ! {
+    // Grab our singleton objects
+    let mut pac = pac::Peripherals::take().unwrap();
```
Comment (thejpster) [inline rp2040-hal-examples/src/bin/spi_eh_bus.rs:104]: I've been trying to remove instances where we have a module and a variable of the same name (`pac`).

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        // UART TX (characters sent from RP2040) on pin 1 (GPIO0)
+        pins.gpio0.into_function(),
+        // UART RX (characters received by RP2040) on pin 2 (GPIO1)
+        pins.gpio1.into_function(),
+    );
+
+    // Set up a uart so we can print out the values from our Spi peripheral
+    let mut uart = hal::uart::UartPeripheral::new(pac.UART0, uart_pins, &mut pac.RESETS)
+        .enable(
+            UartConfig::new(9600.Hz(), DataBits::Eight, None, StopBits::One),
```
Comment (thejpster) [inline rp2040-hal-examples/src/bin/spi_eh_bus.rs:144]: Half our examples run at 9600 and half at 115200. I prefer the latter.

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            clocks.peripheral_clock.freq(),
+        )
+        .unwrap();
+
+    // Set up our SPI pins so they can be used by the SPI driver
+    let spi_mosi = pins.gpio7.into_function::<hal::gpio::FunctionSpi>();
+    let spi_miso = pins.gpio4.into_function::<hal::gpio::FunctionSpi>();
+    let spi_sclk = pins.gpio6.into_function::<hal::gpio::FunctionSpi>();
+    let spi_cs = pins.gpio8.into_push_pull_output_in_state(PinState::High);
+    let spi = hal::spi::Spi::<_, _, _, 8>::new(pac.SPI0, (spi_mosi, spi_miso, spi_sclk));
```
Comment (thejpster) [inline rp2040-hal-examples/src/bin/spi_eh_bus.rs:154]: We should add a comment for this line, to explain what it does.

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        clocks.peripheral_clock.freq(),
+        16.MHz(),
+        embedded_hal::spi::MODE_0,
+    );
+
+    // We are the only task talking to this SPI peripheral, so we can use ExclusiveDevice here.
+    // If we had multiple tasks accessing this, you would need to use a different interface to
+    // ensure that we have exclusive access to this bus (via AtomicDevice or CriticalSectionDevice, for example)
+    // We can safely unwrap here, because the only possible failure is CS assertion failure, but our CS pin is infallible
+    let spi_bus = ExclusiveDevice::new(spi, spi_cs, timer).unwrap();
```
Comment (thejpster) [inline rp2040-hal-examples/src/bin/spi_eh_bus.rs:168]: I love this comment. Excellent.

Comment (thejpster) [reply rp2040-hal-examples/src/bin/spi_eh_bus.rs:168]: But the variable should be called excl_dev or something.

**Review by thejpster — CHANGES_REQUESTED** [review]

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs (hunk 3 above)
Comment (9names) [reply rp2040-hal-examples/src/bin/spi_eh_bus.rs:104]: do you have a preference here? the C programmer in me loves `p`, but maybe `peri` or `peripherals` would be nicer.

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs (hunk 4 above)
Comment (9names) [reply rp2040-hal-examples/src/bin/spi_eh_bus.rs:144]: same.

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs (hunk 5 above)
Comment (9names) [reply rp2040-hal-examples/src/bin/spi_eh_bus.rs:158]: I've updated it to read as below, is this better?
```rust
    // Create new, uninitialized SPI bus with one of the 2 SPI peripherals from our PAC, and our SPI data/clock pins
    let spi = hal::spi::Spi::<_, _, _, 8>::new(pac.SPI0, (spi_mosi, spi_miso, spi_sclk));
```

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs (hunk 1 above)
Comment (9names) [reply rp2040-hal-examples/src/bin/spi_eh_bus.rs:59]: I'm not sure I understand which "this" you're referring to with options like that.

Comment (thejpster) [reply rp2040-hal-examples/src/bin/spi_eh_bus.rs:59]: Sorry. That my fault for using my phone.

Can you change the SPI type parameter to E or ERR?

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs (hunk 3 above)
Comment (thejpster) [reply rp2040-hal-examples/src/bin/spi_eh_bus.rs:104]: The other option is to not import pac and use hal::pac.

**Comment by thejpster** [comment]
I propose some extra comments/updates in https://github.com/rp-rs/rp-hal/tree/spibus_example.

**Comment by 9names** [comment]
Everything in your branch looks fine to me, going to cherry-pick it into here.
I think I'll add the pac->p rename on top though, it tidies up a bit of the initialisation code

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs (hunk 3 above)
Comment (9names) [reply rp2040-hal-examples/src/bin/spi_eh_bus.rs]: Renamed to p *and* use hal::pac

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs (hunk 1 above)
Comment (9names) [reply rp2040-hal-examples/src/bin/spi_eh_bus.rs]: I see where you were going with this, E is an improvement

**Comment by 9names** [comment]
We probably want a 2nd reviewer now that we've both had a hand in here

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+//!
+//! It may need to be adapted to your particular board layout and/or pin
+//! assignment.
+//!
+//! See the top-level `README.md` file for Copyright and license details.
+//!
+//! ## Glossary
+//!
+//! * *SPI Bus* - a shared bus consisting of three shared wires (Clock, COPI and
+//!   CIPO) and a unique 'Chip Select' wire for each device on the bus. An *SPI
```
Comment (jannic) [inline rp2040-hal-examples/src/bin/spi_eh_bus.rs:19]: I agree with avoiding the terms master/slave in technical descriptions. However we still use MOSI/MISO in rp2040-hal and rp2350-hal docs. Shouldn't we use the terms consistently?

**Review by jannic — APPROVED** [review]

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs (hunk 7 above)
Comment (thejpster) [reply rp2040-hal-examples/src/bin/spi_eh_bus.rs:19]: We should be consistent. Using the acronym but not expanding it is perhaps the first step - then renaming it everywhere.

**Review by thejpster — APPROVED** [review]

File: rp2040-hal-examples/src/bin/spi_eh_bus.rs (hunk 7 above)
Comment (thejpster) [reply rp2040-hal-examples/src/bin/spi_eh_bus.rs:19]: Tracking that in https://github.com/rp-rs/rp-hal/issues/846

**Pull request merged**